src/cpp/src/vlm_pipeline.cpp

```diff
@@ -0,0 +1,588 @@
+    ov::Tensor key_padding_mask(ov::element::boolean, {bs, max_patch_len});
+    bool* mask_data = key_padding_mask.data<bool>();
+    size_t embed_len = pipe.m_pos_embed_cache.get_shape().at(2);
+    ov::Tensor pos_embed(ov::element::f32, {max_patch_len, bs, embed_len});  // BLD => L * B * D
+    float* pos_embed_data = pos_embed.data<float>();
+    float* cache_data = pipe.m_pos_embed_cache.data<float>();
+    size_t _d0 = pipe.m_pos_embed_cache.get_shape().at(0);
+    size_t _d1 = pipe.m_pos_embed_cache.get_shape().at(1);
+    for (size_t i = 0; i < bs; ++i) {
+        size_t target_h = target_sizes.at(i).height;
+        size_t target_w = target_sizes.at(i).width;
+        for (size_t h_idx = 0; h_idx < target_h; ++h_idx) {
+            for (size_t w_idx = 0; w_idx < target_w; ++w_idx) {
+                std::copy_n(
+                    cache_data + h_idx * _d1 + w_idx,
+                    embed_len,
+                    pos_embed_data + (h_idx * target_w + w_idx) * bs * embed_len + i * embed_len
+                );
+            }
+        }
+        for (size_t flat = target_h * target_w; flat < max_patch_len; ++flat) {
+            std::fill_n(pos_embed_data + flat * bs * embed_len + i * embed_len, embed_len, 0.0f);
+        }
+        std::fill_n(mask_data + i * max_patch_len, patch_len[i], false);
+        std::fill_n(mask_data + i * max_patch_len + patch_len[i], max_patch_len - patch_len[i], true);
+    }
+    pipe.m_resampler.set_tensor("x", encoded_image);  // [N, H*W, old_hidden_size]
+    pipe.m_resampler.set_tensor("pos_embed", pos_embed);  // [H*W, N, new_hidden_size]
+    pipe.m_resampler.set_tensor("key_padding_mask", key_padding_mask);  // [N, H*W]
+    pipe.m_resampler.infer();
+    return pipe.m_resampler.get_output_tensor();  // [N, query_num, new_hidden_size]
+}
+}
+
+class ov::genai::VLMPipeline::VLMPipelineImpl {
+};
+
+VLMPipeline::VLMPipeline(
+    const std::filesystem::path& model_dir,
+    const Tokenizer& tokenizer,
+    const std::string& device,
+    const ov::AnyMap device_config,
+    ov::Core core
+) :
+    m_vlm_config{
+        utils::from_config_json_if_exists<ov::genai::VLMConfig>(
+            model_dir, "config.json"
+        )
+    },
+    m_tokenizer{tokenizer},
+    m_vision_encoder(model_dir, device, device_config, core),
+    m_resampler{core.compile_model(
+        model_dir / "resampler.xml", device, device_config
+    ).create_infer_request()},
+    m_embedding{core.compile_model(
+        model_dir / "embed_tokens.xml", device, device_config
+    ).create_infer_request()},
+    m_language{core.compile_model(
+        model_dir / "language_model.xml", device, device_config
+    ).create_infer_request()},
+    m_pos_embed_cache{
+        get_2d_sincos_pos_embed(m_vlm_config.hidden_size, {70, 70})
+    },
+    m_is_chat_conversation{false} {
+        m_language.get_tensor("attention_mask").set_shape({1, 0});
+    }
+
+ov::genai::VLMPipeline::~VLMPipeline() = default;
```

Comment:
let's drop it at all

Reply:
It's not possible. Compiler would try to "define" the destructor in header, but there's no info about `VLMPipelineImpl` in the header. `LLMPipeline` follows the same approach.

The reset is in https://github.com/Wovchena/openvino.genai-public/pull/68 except for moving files to `src` and pimpl.